samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -161,72 +119,96 @@ class KafkaSystemProducer(systemName: String,
-        while(!sourceData.latestFuture.isDone && sourceData.exceptionInCallback.get() == null) {
-          try {
-            sourceData.latestFuture.get()
-          } catch {
-            case t: Throwable => error(t.getMessage, t)
-          }
-        }
+      // Only throws InterruptedException, all other errors are handled in send() callbacks
+      producer.flush()
+
+      // Invariant: At this point either
+      // 1. the producer is fine and there are no exceptions to handle   OR
+      // 2. the producer is closed and one or more sources have exceptions to handle
+      //   2a. all new sends get a ProducerClosedException or IllegalStateException (depending on kafka version)
+      //   2b. there are no messages in flight because the producer is closed
+
+      // Flush can be called concurrently for different sources, so any modification of shared objects must be threadsafe
+
+      // We must check for an exception AFTER flush() because when flush() returns all callbacks for messages sent
+      // in that flush() are guaranteed to have completed and we update the exception in the callback.
+      // If there is an exception, we rethrow it here to prevent the checkpoint.
+      val exception = fatalException.get()
+      if (exception != null) {
+        metrics.flushFailed.inc
+        throw new SystemProducerException("Flush failed. One or more batches of messages were not sent!", exception)
+      }
+      trace("Flushed %s." format source)
+    }
+  }
 
-        //if there is an exception thrown from the previous callbacks just before flush, we have to fail the container
-        if (sourceData.exceptionInCallback.get() != null) {
-          metrics.flushFailed.inc
-          throw sourceData.exceptionInCallback.get()
-        } else {
-          trace("Flushed %s." format (source))
+
+  private def handleSendException(currentProducer: Producer[Array[Byte], Array[Byte]], producerException: SystemProducerException, isFatalException: Boolean) = {
+    metrics.sendFailed.inc
+    error(producerException)
+    // The SystemProducer contract is synchronous, so there's no way for us to guarantee that an exception will
+    // be handled by the Task before we recreate the producer, and if it isn't handled, a concurrent send() from another
+    // source could send on the new producer before handling its own exceptions and produce out of order messages.
+    // So we have to handle it right here in the SystemProducer or never recreate the producer.
+    if (dropProducerExceptions) {
+      warn("Ignoring producer exception. All messages in the failed produce request will be dropped!", producerException)
+
+      if (isFatalException) {
+        producerLock.synchronized {
+          // Prevent each callback from recreating producer for the same failed event.
+          if (currentProducer == producer) {
+            info("Creating a new producer for system %s." format systemName)
+            producer = getProducer()
+          }
```

Comment:
Do we need to close currentProducer after recreating a new one, or are we intentionally avoiding it to prevent some out of order delivery scenarios?